core/tee/tee_sql_fs.c

```diff
@@ -0,0 +1,879 @@
+		tee_fs_off_t off;
+
+		if (last_block < old_last_block) {
+			off = block_pos_raw(last_block);
+			rc = sql_fs_ftruncate_rpc(fd, off);
+			if (rc < 0)
+				goto exit;
+		}
+	} else {
+		/* Extend file with zeroes */
+		tee_fs_off_t off = old_length % BLOCK_SIZE;
+		size_t bnum = block_num(old_length);
+		size_t end_bnum = block_num(new_length);
+
+		buf = calloc(1, BLOCK_SIZE);
+		if (!buf) {
+			*errno = TEE_ERROR_OUT_OF_MEMORY;
+			goto exit;
+		}
+
+		while (bnum <= end_bnum) {
+			size_t len = (size_t)BLOCK_SIZE - (size_t)off;
+
+			rc = write_block_partial(errno, fdp, bnum, buf, len,
+						 off);
+			if (rc < 0)
+				goto exit;
+			off = 0;
+			bnum++;
+		}
+	}
+
+	fdp->meta.length = new_length;
+	rc = write_meta(errno, fdp);
+
+exit:
+	sql_fs_end_transaction_rpc(rc < 0);
+	free(buf);
+exit_ret:
+	DMSG("...%d", rc);
+	return rc;
+}
+
+static tee_fs_off_t sql_fs_lseek(TEE_Result *errno, int fd,
+				 tee_fs_off_t offset, int whence)
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	tee_fs_off_t ret = -1;
+	tee_fs_off_t raw_pos;
+	tee_fs_off_t pos;
+
+	DMSG("(fd: %d, offset: %" PRId64 ", whence: %d)...", fd, offset,
+	     whence);
+
+	if (!fdp)
+		goto exit_ret;
+
+	*errno = TEE_ERROR_GENERIC;
+
+	sql_fs_begin_transaction_rpc();
+
+	switch (whence) {
+	case TEE_FS_SEEK_SET:
+		pos = offset;
+		break;
+
+	case TEE_FS_SEEK_CUR:
+		pos = fdp->pos + offset;
+		break;
+
+	case TEE_FS_SEEK_END:
+		pos = fdp->meta.length + offset;
+		break;
+
+	default:
+		*errno = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	raw_pos = pos_to_raw(pos);
+	if (raw_pos < 0)
+		goto exit;
+	raw_pos = sql_fs_lseek_rpc(fd, raw_pos, TEE_FS_SEEK_SET);
+	if (raw_pos < 0)
+		goto exit;
+	ret = raw_to_pos(raw_pos);
+	if (ret < 0)
+		goto exit;
+	*errno = TEE_SUCCESS;
+exit:
+	sql_fs_end_transaction_rpc(ret < 0);
+	fdp->pos = ret;
+exit_ret:
+	DMSG("...%" PRId64, ret);
+	return ret;
+}
+
+static int sql_fs_close(int fd)
+{
+	struct sql_fs_fd *fdp = handle_lookup(&fs_db, fd);
+	int ret = -1;
+
+	DMSG("(fd: %d)...", fd);
+
+	if (!fdp)
+		goto exit;
+
+	sql_fs_close_rpc(fd);
+	put_fdp(fdp);
+
+	ret = 0;
+exit:
+	DMSG("...%d", ret);
+	return ret;
+}
+
+static int sql_fs_open(TEE_Result *errno, const char *file, int flags, ...)
+{
+	struct sql_fs_fd *fdp = NULL;
+	int rflags = flags | TEE_FS_O_RDWR; /* Need to read/write meta */
+	int exists;
+	int fd = -1;
+
+	DMSG("(file: %s, flags: %d)...", file, flags);
+
+	exists = !sql_fs_access_rpc(file, TEE_FS_F_OK);
+	if (flags & TEE_FS_O_CREATE) {
+		if ((flags & TEE_FS_O_EXCL) && exists) {
+			*errno = TEE_ERROR_ACCESS_CONFLICT;
+			goto exit;
+		}
+	} else {
+		if (!exists) {
+			*errno = TEE_ERROR_ITEM_NOT_FOUND;
+			goto exit;
+		}
+	}
+
+	fdp = (struct sql_fs_fd *)calloc(1, sizeof(*fdp));
+	if (!fdp) {
+		*errno = TEE_ERROR_OUT_OF_MEMORY;
+		goto exit;
+	}
+
+	fdp->fd = sql_fs_open_rpc(errno, file, rflags);
+	if (fdp->fd < 0)
+		goto exit;
+
+	fdp->flags = flags;
+
+	fd = read_meta(errno, fdp);
+	if (fd < 0)
```

Comment:
`read_meta` can return several positive values. Seems only rc=meta size if a valid one in case operation succeed.
If so, it should be checked (maybe from `read_meta()`, returning positive or null only for expected meta sizes read.